### PR TITLE
Implement pre-flighting

### PR DESCRIPTION
**Pull request by msirringhaus**

### CODE REVIEW

File: src/crypto/mod.rs
Comment (jschanck) [inline src/crypto/mod.rs]: I'd prefer to store just the protocol and the permissions. If we ever need to recover the `PinUvAuthToken` from a `PinUvAuthParam` we've done something wrong. If the state machine needs to execute a sequence of commands with different `PinUvAuthParam`s derived from the same `PinUvAuthToken` then it should cache the `PinUvAuthToken` itself.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: This should be fast, so I'm not too concerned about it being cancellable.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: Let's make `dev` the first argument for consistency with other functions.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: The spec is not very clearly written, but I don't think it's ambiguous. IIUC, it's saying that you only need to perform this `CheckKeyHandle` step when `up=true` and the transport is USB. The `CheckKeyHandle` step should always use `0x07`.

In the `up=false` case you can just iterate over the allow list and send `U2F_AUTHENTICATE` messages with the control byte equal to `0x08`, as this is non-blocking. Likewise for NFC with `up=true` you can just iterate over the allow list and use control byte `0x03` as this is also non-blocking.

It should be fine to perform `CheckKeyHandle` in the `up=false` case, as you're doing here, but it's not strictly necessary.

File: src/ctap2/commands/get_assertion.rs
Comment (jschanck) [inline src/ctap2/commands/get_assertion.rs:335]: Github won't let me comment on the right place, but note that we should be setting flags=`0x08` in the `up=false` case here.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: I think "pre-flighting" is more general than "filtering a credential id list". You can also pre-flight to determine which authenticators to blink, and you can pre-flight to get a list of available credentials for the WebAuthn conditional UI. So this needs another name.

I didn't like that we were sending messages to the device as part of `ctap1_format`, so the approach here is an improvement.

Personally, I wouldn't implement this as a trait. I would move `do_pre_flight_ctap2` to `statemachine.rs` (and rename it), and I would have it take a credential ID list and return a filtered list. I'd then put the logic for interpreting that filtered list in-line in each of `StateMachine::register` and `StateMachine::sign`.

As I mentioned in another comment, we have to be careful in `StateMachine::sign` as the filtered credential ID list can still exceed the maximum list size.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: I think we want to update `GetAssertion::new` to take an `Option<PinUvAuthParam>`.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: I don't think this is right for `GetAssertion`. An allow list might have a number of valid credentials that exceeds the chunk size. This will not return all of them. Changing this to append the chunk won't work because then the final request won't fit in a single chunk.

File: src/crypto/mod.rs
Comment (msirringhaus) [reply src/crypto/mod.rs]: Esp. for "CredentialManagement" and "BioEnrollment", we will execute a sequence of commands. But I'm fine with letting them cache the `PinUvAuthToken` on their own. Looked just simpler to have the `PinUvAuthParam` cache it for us, and its a bit of a duplication of protocol and permissions.

File: src/ctap2/preflight.rs
Comment (msirringhaus) [reply src/ctap2/preflight.rs]: Ok, yes it makes sense. I do see that we could skip the filtering-step here for `up=false`, but that would be another major refactoring to actually implement properly. And I don't really see `up=false` being used in practice, so I think its ok to send one additional request to the token in that case, that ideally could be skipped.

I'll update the comment accordingly, and actually also fix the serialization of GetAssertion, because that sets the flag to zero, in case of `up=false`, which is explicitly forbidden by the CTAP1-spec. (I honestly only copied the existing code here, and didn't double-check everything).

File: src/ctap2/commands/get_assertion.rs
Comment (msirringhaus) [reply src/ctap2/commands/get_assertion.rs:335]: Yes, indeed.

File: src/ctap2/preflight.rs
Comment (msirringhaus) [reply src/ctap2/preflight.rs]: You don't like traits, do you? :smile: 
In general, I would like to leave the logic/function in `preflight.rs`, because `statemachine.rs` is getting very large (it's actually already way too large for my taste). Having all the functionality inside StateMachine also makes it harder to test the individual pieces. But in this case it's mostly fine, so I'll refactor.

Regarding the name, the spec defines "pre-flight" as:
> pre-flight
>
>  In order to determine whether authenticatorMakeCredential's excludeList or authenticatorGetAssertion's allowList contain credential IDs that are already present on an authenticator, a platform typically invokes authenticatorGetAssertion with the "up" option key set to false and optionally pinUvAuthParam one or more times. 

so I think the name is correct? But I will rename it to `do_credential_list_filtering_*()` nonetheless, to make it easier to follow without knowing all of the spec.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: Let's resolve this TODO before landing this.

File: src/statemachine.rs
Comment (jschanck) [inline src/statemachine.rs]: I think we can just request both permissions. They're automatically granted when using internal UV. The token can disallow *both* MC and GA when using PIN based UV, but it can't disallow just one of the two.

File: src/statemachine.rs
Comment (jschanck) [inline src/statemachine.rs:852]: If the input `allow_list` is non-empty and the output `allow_list` is empty, then we should exit here. We don't want the token to interpret the empty `allow_list` as a request for discoverable credentials.

File: src/statemachine.rs
Comment (jschanck) [inline src/statemachine.rs:589]: I don't think `unwrap_result` improves readability here. Better to explicitly handle the cases, especially since we can return early with `CredentialExcluded` if the resulting `exclude_list` is non-empty.

File: src/statemachine.rs
Comment (jschanck) [inline src/statemachine.rs]: `dummy_make_credentials_cmd` can't actually fail here. We should just hard-code a bogus client data hash so that we don't need to handle a failure in the crypto backend.

File: src/ctap2/preflight.rs
Comment (jschanck) [inline src/ctap2/preflight.rs]: We probably don't need this function, see my comment on its use below.

File: src/ctap2/preflight.rs
Comment (jschanck) [reply src/ctap2/preflight.rs]: I was convinced by your message on matrix that this is OK. The spec does not require us to present all of the credentials in the allow list to the authenticator.

File: src/statemachine.rs
Comment (msirringhaus) [reply src/statemachine.rs]: I had it because the spec says:
> Note: The platform SHOULD request only the permissions absolutely necessary.

 But I think you are right, we can probably just always request MC and GA here.

File: src/statemachine.rs
Comment (msirringhaus) [reply src/statemachine.rs:589]: I'm not 100% sure why yet, but Chromium is explicitly requiring a touch by the user in both error-cases (`CredentialExcluded` here and `NoCredentials` for GetAssertion). For GetAssertion, it even creates a dummy request, just to get the user-interaction. I'm still skimming the spec to find out, if it is mentioned there somewhere.

Either way, we are currently inconsistent (user-interaction for MC, no user-interaction for GE). We should decide which way to go, and do it consistently. Both ways are easily possible.

Comment (msirringhaus) [reply src/statemachine.rs:589]: Ah, [there](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-20210615.html#op-makecred-step-if-excludeList) it is for MC:
> Note: A user presence test is required for CTAP2 authenticators, before the RP is told that the authenticator is already registered, to behave similarly to CTAP1/U2F authenticators.

I currently can't find a strict requirement for GA, though. But we did this in the U2F-case as well: https://github.com/mozilla/authenticator-rs/blob/main/src/statemachine.rs#L229-L238

File: src/statemachine.rs
Comment (msirringhaus) [reply src/statemachine.rs]: Good idea!

File: src/statemachine.rs
Comment (msirringhaus) [reply src/statemachine.rs:852]: Yes, I added the check + a dummy command for getting some user presence test (see the other comment).

File: src/ctap2/preflight.rs
Comment (msirringhaus) [reply src/ctap2/preflight.rs]: Ok, I think I understand this now.
Neither `maxCredentialIdLength` nor `maxCredentialCountInList` are allowed to be zero according to spec ("MUST be greater than zero if present."). So if the device returns 0 here, it can't be trusted so to speak. So a fallback to the bare minimum of functionality is advisable.

File: src/ctap2/preflight.rs
Comment (msirringhaus) [reply src/ctap2/preflight.rs]: I honestly don't really see a need for that.

**Review by jschanck — APPROVED** [review]
Looks good, thanks!

File: src/statemachine.rs
Comment (jschanck) [reply src/statemachine.rs:589]: Ah, right, good point!

**Pull request merged**
